### DISCUSSION TITLE
Cause for Resets (PyBoard)

### DISCUSSION THREAD

**Discussion by Davis-Garrad** [micropython/micropython#11084] [Other]
Hello,

I'm working on a PyBoard right now in a relatively complex temperature control system. A PyBoard accepts commands from a controlling computer through UART and all functionality works well, until it doesn't. Seemingly randomly, the PyBoard itself will soft restart - this is an issue because this control is very sensitive. I've looked around a bit, but I couldn't find any documentation for how one can determine the cause of a reset in a PyBoard - is this due to the power supply not providing enough current? Maybe a PWM timer issue (as has happened in the past)? Something else?

I'd love to know if the reason for an automatic reset is logged anywhere, or if I could configure the PyBoard to do as such. Edit: (Separate from the machine.reset_cause() function - I know of this, but it is too vague. I don't run any watchdog timers (unless one already exists on the PyBoard without initialization) and it won't really help to know if it is a soft or hard reset I don't think) 

Thanks in advance,
Davis

**Comment by davefes** [comment]
Maybe a software watchdog.  I use one to log the time and a message (that it was a wdt) on any time-outs.  In suspect routines using a try/except maybe you could get more information about which area of code is causing the problem, when it times-out.

**Reply by Davis-Garrad** [reply]
Well, here's the thing: The PyBoard is just running a while loop that accepts commands - it doesn't reset during regular operations, nor is the loop hanging. As far as I'm aware, the code runs perfectly fine. I was just looking for a way to tell if the reset was due to hardware, and if so which part. I've had issues in the past where using timers for PWM has caused resets and I'm sure unstable power sources will lead to bad behaviour, etc. so it would be helpful to know which part of the hardware is failing. Thanks for the suggestion though!

**Reply by Josverl** [reply]
Have you monitored heap space/fragmentation?
You could write mem_free output to a 2nd uart and store/analyze that for clues
